ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -110,57 +210,96 @@ data Query blk result where
 
-instance (ShowQuery (BlockQuery blk), StandardHash blk) => ShowQuery (Query blk) where
+instance (forall fp. ShowQuery (BlockQuery blk fp), StandardHash blk) => ShowQuery (Query blk) where
   showResult (BlockQuery blockQuery) = showResult blockQuery
   showResult GetSystemStart          = show
   showResult GetChainBlockNo         = show
   showResult GetChainPoint           = show
 
-instance Eq (SomeSecond BlockQuery blk) => Eq (SomeSecond Query blk) where
-  SomeSecond (BlockQuery blockQueryA) == SomeSecond (BlockQuery blockQueryB)
-    = SomeSecond blockQueryA == SomeSecond blockQueryB
-  SomeSecond (BlockQuery _) == _ = False
+instance Show (SomeBlockQuery (BlockQuery blk)) => Show (Some (Query blk)) where
+  show (Some (BlockQuery blockQueryA))  = "Query " ++ show (SomeBlockQuery blockQueryA)
+  show (Some GetSystemStart)            = "Query GetSystemStart"
+  show (Some GetChainBlockNo)           = "Query GetChainBlockNo"
+  show (Some GetChainPoint)             = "Query GetChainPoint"
 
-  SomeSecond GetSystemStart == SomeSecond GetSystemStart = True
-  SomeSecond GetSystemStart == _                         = False
+------
+-- Eq
+------
 
-  SomeSecond GetChainBlockNo == SomeSecond GetChainBlockNo  = True
-  SomeSecond GetChainBlockNo == _                           = False
-
-  SomeSecond GetChainPoint == SomeSecond GetChainPoint  = True
-  SomeSecond GetChainPoint == _                         = False
-
-instance Show (SomeSecond BlockQuery blk) => Show (SomeSecond Query blk) where
-  show (SomeSecond (BlockQuery blockQueryA))  = "Query " ++ show (SomeSecond blockQueryA)
-  show (SomeSecond GetSystemStart)            = "Query GetSystemStart"
-  show (SomeSecond GetChainBlockNo)           = "Query GetChainBlockNo"
-  show (SomeSecond GetChainPoint)             = "Query GetChainPoint"
+instance SameDepIndex1 (BlockQuery blk) => SameDepIndex (Query blk) where
+  sameDepIndex (BlockQuery blockQueryA) (BlockQuery blockQueryB)
+    = (\Refl -> Refl) <$> sameDepIndex1 blockQueryA blockQueryB
+  sameDepIndex (BlockQuery _) _
+    = Nothing
+  sameDepIndex GetSystemStart GetSystemStart
+    = Just Refl
+  sameDepIndex GetSystemStart _
+    = Nothing
+  sameDepIndex GetChainBlockNo GetChainBlockNo
+    = Just Refl
+  sameDepIndex GetChainBlockNo _
+    = Nothing
+  sameDepIndex GetChainPoint GetChainPoint
+    = Just Refl
+  sameDepIndex GetChainPoint _
+    = Nothing
 
+------
+-- Serialization
+------
 
--- | Exception thrown in the encoders
-data QueryEncoderException blk =
-    -- | A query was submitted that is not supported by the given 'QueryVersion'
-    QueryEncoderUnsupportedQuery
-         (SomeSecond Query blk)
-         QueryVersion
+-- | Exception thrown in the encoders: A query was submitted that is not supported by the given 'QueryVersion'
+data QueryEncoderException = forall blk. Show (Some (Query blk)) =>
+    QueryEncoderUnsupportedQuery (Some (Query blk)) QueryVersion
```

Comment:
Is hiding `blk` as an existential necessary?

Reply:
It is useful to not require `Typeable blk` in some places. Exceptions are better if they are monomorphic